### PR TITLE
[`flake8-bugbear`] Implement `quoted-fstring-value` (`B907`)

### PR DESCRIPTION
**Pull request by alexcdennis**
## Summary

Implements rule B907 `quoted-fstring-value`.

Tracking Issue: #3758 

From the [bugbear documentation](https://github.com/PyCQA/flake8-bugbear/tree/18abb9fdec66430d36010ffcee2d7dc54c802d3a?tab=readme-ov-file#opinionated-warnings)
> B907: Consider replacing f"'{foo}'" with f"{foo!r}" which is both easier to read and will escape quotes inside foo if that would appear. The check tries to filter out any format specs that are invalid together with !r. If you're using other conversion flags then e.g. f"'{foo!a}'" can be replaced with f"{ascii(foo)!r}". Not currently implemented for python<3.8 or str.format() calls.

## Changes from original

The original rule in bugbear was marked as optional/opinionated (https://github.com/PyCQA/flake8-bugbear/pull/333) because there were several complaints about limitations (https://github.com/PyCQA/flake8-bugbear/issues/329, https://github.com/PyCQA/flake8-bugbear/issues/331, https://github.com/PyCQA/flake8-bugbear/issues/332, https://github.com/astral-sh/ruff/issues/1893#issuecomment-1383132618).

The cause of all these limitations was that the suggested fix (to always add `!r`) was too opinionated and could changed the semantics of the output.

As such, this implementation diverges from the original by changing the framing in the documentation and error messages to remove all reference to the problematic suggested fix. Instead, it opts for a more nuanced approach which requires the user to consider their own fix on a case-by-case basis, while also providing an opt out (the user may add a noop to the code).

## Test Plan

`cargo nextest run`

## Sources

Implementation inspired by [original bugbear implementation](https://github.com/PyCQA/flake8-bugbear/blob/18abb9fdec66430d36010ffcee2d7dc54c802d3a/bugbear.py#L1419)
Test cases include all test cases from the [original bugbear test file](https://github.com/PyCQA/flake8-bugbear/blob/18abb9fdec66430d36010ffcee2d7dc54c802d3a/tests/b907.py), plus extras

### ruff
#1893
https://github.com/astral-sh/ruff/issues/2954#issuecomment-1483594606

### bugbear
https://github.com/PyCQA/flake8-bugbear/issues/319

### CODE REVIEW

**Comment by charliermarsh** [comment]
It seems like the check in bugbear suffered from a lot of false positives (and hence was moved to the opinionated section). Do you know how many of those are present in the implementation here? What are the limitations of the rule?

**Comment by alexcdennis** [comment]
This rule has all of the same limitations as the original, since it is the same algorithm.

My understanding is that we wanted behavior parity with bugbear, regardless of whether the rule is good or useful.

The limitations are as follows:
This rule may suggest a fix that changes the meaning of the output, specifically if the value is not a string (https://github.com/PyCQA/flake8-bugbear/issues/329, https://github.com/astral-sh/ruff/issues/1893#issuecomment-1383132618) or if the value represents a string where there is a different meaning between double quotes and single quotes (like an SQL query https://github.com/PyCQA/flake8-bugbear/issues/331).
The former cannot be overcome without type inference, and the later cannot be overcome at all because it depends upon the intent of the programmer for the arbitrary semantics of the string.

For this reason, I tried to word the message as a suggestion for the user to consider if this is a problem on a case-by-case basis instead of as something that the user must fix. It may require more rewording to make this clearer (eg. to remove any implication of a fix from the wording of the message).

**Comment by alexcdennis** [comment]
I have changed the framing around the rule by changing the suggested fix. I believe that this overcomes all previously suggested limitations, as discussed above.

Summary:
Instead of the original suggested fix of "always add `!r`", the new suggested fix is "consider on a case-by-case basis; if you want to signal that this is intended behavior, add a noop"

Open question:
Is the new documentation/error message too wordy/nuanced?

**Comment by nidhishgajjar** [comment]
**Orb Code Review** (powered by GLM-4.7 on [Orb Cloud](https://orbcloud.dev))## SummaryI've reviewed the changes in this PR (PR #11977). The diff contains 1350 lines.## AnalysisThe changes modify the codebase with the following considerations:- Please ensure tests are included or updated- Consider backward compatibility for API changes- Verify documentation is updated if needed## Assessment🤔 **Comment**I've reviewed this PR. Please provide more details about:1. What problem this PR solves2. Any breaking changes introduced3. Test coverage for the new code
